requests/utils.py

```diff
@@ -541,14 +628,23 @@ def should_bypass_proxies(url):
                     # to apply the proxies on this URL.
                     return True
 
+
     # If the system proxy settings indicate that this URL should be bypassed,
     # don't proxy.
     # The proxy_bypass function is incredibly buggy on OS X in early versions
     # of Python 2.6, so allow this call to fail. Only catch the specific
     # exceptions we've seen, though: this call failing in other ways can reveal
     # legitimate problems.
+    #
+    # Additionally, the proxy_bypass function might cause unexpected delay when
+    # it called socket.gethostbyaddr internally (refer to #2988). Thus we cache
+    # the function call. Also, to invalidate the cache from time to time (as we
+    # don't know when to expire the cache exactly), we use the TTLCache
+    # mechanism here.
     try:
-        bypass = proxy_bypass(netloc)
+        bypass = proxy_bypass_cacher.get(netloc)
+        if bypass is None:
+            bypass = proxy_bypass_cacher.set(netloc, proxy_bypass(netloc))
```

Comment:
`set` doesn't return the value so this won't work.

    bypass = proxy_bypass(netloc)
    proxy_bypass_cacher.set(netloc, bypass)